### PR TITLE
[FLINK-7533] Let LeaderGatewayRetriever retry failed gateway retrievals

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

In case that the `LeaderGatewayRetriever` fails to retrieve the leader gateway, it should retry the retrieval upon the next request. Otherwise, a once failed retrieval will only be resolved if the leader address changes.

## Verifying this change

Adds the test case `LeaderGatewayRetrieverTest#testGatewayRetrievalFailures`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by tillrohrmann** [comment]
Rebasing onto the latest master.

File: flink-core/src/main/java/org/apache/flink/util/OptionalConsumer.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Utility class which allows to run code depending on whether the
+ * optional has a value or is empty.
+ *
+ * <p>This code has been copied from: https://stackoverflow.com/a/29395324/4815083.
+ *
+ * @param <T> type of the optional
+ */
+public class OptionalConsumer<T> {
+	private Optional<T> optional;
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/OptionalConsumer.java]: could be final

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java [hunk 2]
```diff
@@ -137,7 +138,6 @@ public TaskManagerLogHandler(
 		this.blobView = Preconditions.checkNotNull(blobView, "blobView");
 	}
 
-	@Override
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java]: revert as it implements WebHandler

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebHandler.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+package org.apache.flink.runtime.webmonitor;
+
+/**
+ * Marker interface for web handlers which can describe their paths.
+ */
+public interface WebHandler {
+
+	/**
+	 * Paths to register the handler under.
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebHandler.java]: why not copy the existing javadocs from RuntimeMonitorHandlerBase?

File: flink-core/src/main/java/org/apache/flink/util/OptionalConsumer.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-core/src/main/java/org/apache/flink/util/OptionalConsumer.java]: true, will change it.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java]: True, will change it.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebHandler.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebHandler.java]: Will change it.

**Pull request closed without merging**